### PR TITLE
introduce an operation filter for all modrules

### PR DESCRIPTION
**Pull request by orbatschow**
This will fix https://github.com/kubemod/kubemod/issues/103

I haven't extended the unit tests, as they currently do not fit the new feature. As far as I can see
all tests just simulate a "CREATE" opeation. I would be willing to refactor the tests, to use go table tests and a more dynamic approach,
but I would love to get some feedback on this first.

The PR will also make the telepresence helper comply with the latest telepresence version and bump go to 1.18

### CODE REVIEW

**Comment by orbatschow** [comment]
The run seems to fail, I'm not quite sure whether this is related to the go version of the VM
Simply running the `make test` command locally works.
I have bumped the go versions, maybe this already fixes the issues.

**Comment by vassilvk** [comment]
Thanks @orbatschow!
Hopefully I'll have some time to look into this PR during my off hours this week, if not, then this coming weekend.

In the meantime, can you please bump golang to 1.18 in the remaining places where 1.17 is being used?
Those would be `Dockerfile` and `Makefile`.

File: core/modrulestore.go [hunk 1]
```diff
@@ -238,7 +263,7 @@ func extractLastAppliedConfiguration(jsonv interface{}) []byte {
 
 // CalculatePatch calculates the set of patch operations to apply against a given resource
 // based on the ModRules matching the resource.
-func (s *ModRuleStore) CalculatePatch(namespace string, originalJSON []byte, operationLog logr.Logger) (interface{}, []ctrljsonpatch.JsonPatchOperation, error) {
+func (s *ModRuleStore) CalculatePatch(operation v1beta1.ModRuleOperation, namespace string, originalJSON []byte, operationLog logr.Logger) (interface{}, []ctrljsonpatch.JsonPatchOperation, error) {
```
Comment (vassilvk) [inline core/modrulestore.go:266]: Please rename `operation` to `admissionOperation`.

File: core/modrulestore.go [hunk 2]
```diff
@@ -333,7 +358,7 @@ func (s *ModRuleStore) CalculatePatch(namespace string, originalJSON []byte, ope
 }
 
 // DetermineRejections checks if the given object should be rejected based on the current Reject ModRules stored in the namespace.
-func (s *ModRuleStore) DetermineRejections(namespace string, jsonv interface{}, operationLog logr.Logger) []string {
+func (s *ModRuleStore) DetermineRejections(operation v1beta1.ModRuleOperation, namespace string, jsonv interface{}, operationLog logr.Logger) []string {
```
Comment (vassilvk) [inline core/modrulestore.go:361]: Please rename `operation` to `admissionOperation`.

File: core/modrulestore.go [hunk 3]
```diff
@@ -140,14 +141,22 @@ func (s *ModRuleStore) Delete(namespace string, name string) {
...
 	var potentialRules []*ModRuleStoreItem
 
 	s.rwLock.RLock()
 	defer s.rwLock.RUnlock()
 
 	processPotentialRule := func(mrsi *ModRuleStoreItem) {
+
+		// if there is an operation defined on the ModRule, check if it matches the admission request operation
+		if len(mrsi.modRule.Spec.Operation) != 0 {
+			if !slices.Contains(mrsi.modRule.Spec.Operation, operation) {
```
[the remark this answers is not in the document]
Comment (vassilvk) [reply core/modrulestore.go:155]: Please don't use experimental package `golang.org/x/exp/slices`.
The code in it is unreliable (as per its own documentation).
Implementing `contains` logic in go is trivial. The idiomatic way would be to do it inline.
Example: https://github.com/kubemod/kubemod/blob/master/core/modrulestoreitem.go#L520

File: go.mod [hunk 4]
```diff
@@ -18,9 +17,75 @@ require (
 	github.com/onsi/gomega v1.10.2
 	github.com/segmentio/ksuid v1.0.3
 	go.uber.org/zap v1.10.0
+	golang.org/x/exp v0.0.0-20221031165847-c99f073a8326
```
Comment (vassilvk) [inline go.mod:20]: Please see comment about not using `golang.org/x/exp/slices` in `modrulestore.go`.

File: samples/modrules/modrule-7.yaml [hunk 5]
```diff
@@ -0,0 +1,17 @@
+apiVersion: api.kubemod.io/v1beta1
+kind: ModRule
+metadata:
+  name: reject-namespace-deletion
+  namespace: kubemod-system
+spec:
+  type: Reject
+  operation:
```
Comment (vassilvk) [inline samples/modrules/modrule-7.yaml:8]: Please rename to `admissionOperations`.

**Review by vassilvk — CHANGES_REQUESTED** [review]
Thank you @orbatschow!
I'm done with my review. Please see comments.

**Comment by orbatschow** [comment]
@vassilvk 

I will resolve all those conflicts, but one topic aside from that:

The CRD contains fields like `match` and `patch` which are also arrays, but written singular. I was a bit confused, but thought, that doing the same for `admissionOperation` would at least be consistent with the other properties. Should be rename `operation` to `admissionOperations` or be consistent with the other properties and name it `admissionOperation`?

**Comment by vassilvk** [comment]
Hi @orbatschow,

I hear you. If you think of `match` and `patch` as nouns this makes sense.

That said, in the context of a ModRule, `match` and `patch` are verbs :)
It's just natural to think of the two sections as "**Match** this resource against these conditions, and if it matches, then **patch** it". So in the documentation, I refer to `match` and `patch` as the **match section** and the **patch section**, not the matches and the patches.

The semantics of `admisionOperations` are closer to the ones of `matchValues` (which you can find as a property of the match criteria item object).

**Comment by orbatschow** [comment]
@vassilvk 

The naming now makes sense to me, I missed that.

I tried to resolve all other comments, please have a look at the `potentialRules` matching, i think it should be a lot better now and also I think this is the best place to match the rule.

**Comment by vassilvk** [comment]
Hi @orbatschow - thank you!
I have not looked at the update yet, but before I get to it, can you please also make the following change:

In order to maintain backward compatibility with previous versions of KubeMod, we need to make sure that when `admissionOperations` is not specified for a ModRule, it only applies to `CREATE` and `UPDATE` operations.
Currently ModRules with no `admissionOperations` trigger for `DELETE` as well.
If we keep the current behavior, then all ModRules people have created in the past, will also be triggered for `DELETE` operations, which would unnecessarily slow down KubeMod execution for DELETE operations.

In addition, I don't see many use cases where a ModRule would apply to `CREATE/UPDATE`, but also `DELETE`.
A typical ModRule would either apply to `CREATE/UPDATE` or `DELETE` operations, but not all three.

**Comment by orbatschow** [comment]
@vassilvk 

I have reverted these changes.

**Comment by vassilvk** [comment]
Hi @orbatschow, thanks for your work!
Hopefully I’ll have some time to review this week. If not, I’ll take a look over the weekend.

**Review by vassilvk — APPROVED** [review]

**Comment by vassilvk** [comment]
Hi @orbatschow,

Than you for your PR. Instead of going back and forth with reviews, I chose to [submit a commit](https://github.com/kubemod/kubemod/pull/104/commits/995931a591e539318d9cac6360b08fca0d84170d) into your pull request.

The main changes:
- It's best to not ask users to manually add DELETE to the admission webhook configuration in order to unlock your use case. I added it to the list of target operations.
- I made sure that admissionOperations defaults to `CREATE` and `UPDATE` in order to maintain backward compatibility with the previous default behavior of the operator
- I moved the logic of filtering by operation to the stage prior to the matching, in order to make sure modrules that do not apply to the current operation, do not interfere with the execution tier

I have one ask of you - can you please review my commit and deploy/test on your end all the modrules in `./samples/modrules` by deploying them, and then running `kubectl apply -f ./samples/stack`. You should see several log lines indicating that the modrules did modify the target resources.

**Comment by orbatschow** [comment]
Hey @vassilvk 

I tested your changes, they seem to be working fine.

test:

```
…bat in …/github.com/kubemod/kubemod on  master [$!] on ⛵default via  v1.19.1 on ☁️  (RegionOne)
[I] ➜  k apply -f samples/modrules/
modrule.api.kubemod.io/modrule-1 created
modrule.api.kubemod.io/modrule-2 created
modrule.api.kubemod.io/reject-malicious-services created
modrule.api.kubemod.io/patch-istio-namespace created
modrule.api.kubemod.io/patch-clusterrole created
modrule.api.kubemod.io/patch-my-namespace-pod created
modrule.api.kubemod.io/reject-namespace-deletion created
…bat in …/github.com/kubemod/kubemod on  master [$!] on ⛵default via  v1.19.1 on ☁️  (RegionOne)
[I] ➜  k apply -f samples/stack/
clusterrole.rbac.authorization.k8s.io/my-clusterrole configured
namespace/my-namespace configured
deployment.apps/nginx unchanged
deployment.apps/nginx configured
service/nginx unchanged
…bat in …/github.com/kubemod/kubemod on  master [$!] on ⛵default via  v1.19.1 on ☁️  (RegionOne)
[I] ➜
```

logs:

```
{"level":"info","ts":"2022-12-14 15:44:57.642+0100","logger":"modrule-webhook","msg":"Applying ModRule patch","request uid":"d40abd25-0fde-41d4-b288-251661cbd420","namespace":"","resource":"clusterroles/my-clusterrole","operation":"CREATE","patch":[{"op":"replace","path":"/metadata/annotations/kubectl.kubernetes.io~1last-applied-configuration","value":"{\"apiVersion\":\"rbac.authorization.k8s.io/v1\",\"kind\":\"ClusterRole\",\"metadata\":{\"annotations\":{},\"labels\":{\"color\":\"blue\"},\"name\":\"my-clusterrole\"}}"},{"op":"add","path":"/metadata/labels","value":{"color":"blue"}}]}
{"level":"info","ts":"2022-12-14 15:44:57.721+0100","logger":"modrule-webhook","msg":"Applying ModRule patch","request uid":"f64c799b-b34f-45eb-a587-e111ce6438df","namespace":"my-namespace","resource":"namespaces/my-namespace","operation":"CREATE","patch":[{"op":"replace","path":"/metadata/annotations/kubectl.kubernetes.io~1last-applied-configuration","value":"{\"apiVersion\":\"v1\",\"kind\":\"Namespace\",\"metadata\":{\"annotations\":{\"hello\":\"whatever\"},\"labels\":{\"color\":\"blue\",\"flavor\":\"orange\",\"istio.io/rev\":\"canary\"},\"name\":\"my-namespace\"}}"},{"op":"add","path":"/metadata/labels/istio.io~1rev","value":"canary"}]}
{"level":"info","ts":"2022-12-14 15:44:58.015+0100","logger":"modrule-webhook","msg":"Applying ModRule patch","request uid":"4f17d0ec-84ac-4794-9ebe-945c96113db6","namespace":"default","resource":"deployments/nginx","operation":"CREATE","patch":[{"op":"replace","path":"/metadata/annotations/kubectl.kubernetes.io~1last-applied-configuration","value":"{\"apiVersion\":\"apps/v1\",\"kind\":\"Deployment\",\"metadata\":{\"annotations\":{},\"labels\":{\"app\":\"nginx\",\"color\":\"whatever\"},\"name\":\"nginx\",\"namespace\":\"default\"},\"spec\":{\"replicas\":1,\"selector\":{\"matchLabels\":{\"app\":\"nginx\"}},\"template\":{\"metadata\":{\"labels\":{\"app\":\"nginx\"}},\"spec\":{\"containers\":[{\"image\":\"bitnami/nginx:1.14.2\",\"name\":\"nginx\",\"ports\":[{\"containerPort\":8080,\"protocol\":\"TCP\"}],\"resources\":{\"limits\":{\"cpu\":\"500m\",\"memory\":\"1Gi\"}}},{\"command\":[\"sh\",\"-c\",\"while true; do sleep 5; done;\"],\"image\":\"alpine:3\",\"name\":\"injected\"}]}}}}"},{"op":"add","path":"/metadata/labels/color","value":"whatever"},{"op":"add","path":"/spec/template/spec/containers/1","value":{"command":["sh","-c","while true; do sleep 5; done;"],"image":"alpine:3","name":"injected"}},{"op":"replace","path":"/spec/template/spec/containers/0/ports/0/containerPort","value":8080},{"op":"replace","path":"/spec/template/spec/containers/0/image","value":"bitnami/nginx:1.14.2"}]}
{"level":"info","ts":"2022-12-14 15:44:58.021+0100","logger":"modrule-webhook","msg":"Applying ModRule patch","request uid":"bfe0e2a4-831a-46cc-9c3f-023786f6776f","namespace":"my-namespace","resource":"pods/","operation":"CREATE","patch":[{"op":"add","path":"/metadata/labels/flavor","value":"orange"}]}
```

One thing I had to change though is making the `admissionOperations` optional, otherwise all rules will fail to deploy.
I have updated the PR to reflect those changes

**Comment by vassilvk** [comment]
Great - thank you @orbatschow!
And it absolutely makes sense to make the property optional - good catch :)
Note that I pushed one more commit with the manifests regenerated with the original version of controller-gen installed by the Makefile instructions.
I plan to merge with master and release with KubeMod v0.18.0 later tonight.

**Comment by orbatschow** [comment]
@vassilvk thanks alot for the great work, looking forward for the release :partying_face:

**Pull request merged**